cpp/src/arrow/compute/kernels/vector_cumulative_sum.cc

```diff
@@ -0,0 +1,208 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/array/array_base.h"
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/api_vector.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/result.h"
+#include "arrow/visit_type_inline.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+namespace {
+
+struct Add {
+  template <typename T, typename Arg0, typename Arg1>
```

Comment:
How can we reuse the *addition* kernels directly (eg.., `Add::Call(...)`) in both the scalar arithmetic file and this file?
A more general question is: how can we reuse the *arithmetic* kernels across the compute files?

We could move all the addition kernels to a common file and leave the other arithmetic kernels (subtract, multiply, etc.) in `scalar_arithmetic.cc` but this feels strange. I propose to move all the arithmetic kernels to `arithmetic_internal.h` so that basic arithmetic operations can be used consistently in the compute kernels. Then, `scalar_arithmetic.cc` would contain the function registration logic and any other related code.

To be consistent with the Add kernels, we would need to add a safe version (checks for overflows) for the cumulative sum.

Cumulative sum would make use of:
* `Add` (included here) and [`AddChecked`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L142)
* [`AddTimeDuration` variants`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L244)

cc @pitrou